### PR TITLE
Thing properties

### PR DESCRIPTION
**Pull request by tomhoefer**

### CODE REVIEW

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java [hunk 1]
```diff
@@ -233,10 +236,15 @@ public Thing createThing(ThingTypeUID thingTypeUID, Configuration configuration,
         if (thingType != null) {
             Thing thing = ThingFactory.createThing(thingType, thingUID, configuration, bridgeUID,
                     getConfigDescriptionRegistry());
+            Map<ThingProperty, String> thingProperties = getThingProperties(thing);
+            if (thingProperties != null) {
+                for(ThingProperty configProperty : thingProperties.keySet()) {
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java]: missing space between "for" and "(" (using the ESH code formatter / clean up).

**Comment by kaikreuzer** [comment]
Could you please add license headers on all new files?

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java [hunk 2]
```diff
@@ -246,5 +254,14 @@ protected ConfigDescriptionRegistry getConfigDescriptionRegistry() {
...
-
+    
+    /**
+     * Retrieves the {@link ThingProperty}s for the given {@link Thing}. Sub-classes should overwrite this operation
+     * in order to provide these kind of properties for the things to be created.
+     * 
+     * @param thing the {@link Thing} for which the {@link ThingProperty}s are to be retrieved
+     */
+    protected Map<ThingProperty, String> getThingProperties(Thing thing) {
+        return Maps.newHashMap();
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java]: If possible, I would avoid the use of Guava here.

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java [hunk 3]
```diff
@@ -246,5 +254,14 @@ protected ConfigDescriptionRegistry getConfigDescriptionRegistry() {
         }
         return configDescritpionRegistryServiceTracker.getService();
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java]: Could you fix this typo with this PR as well? Thanks :-)

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/ThingProperty.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+/**
+ * The {@link ThingProperty} enumeration defines all base configuration properties for a {@link Thing}. They
+ * will be injected into the {@link Configuration} of a thing during its instantiation (compare
+ * {@link BaseThingHandlerFactory#createThing(ThingTypeUID, Configuration, ThingUID, ThingUID)}). For this purpose
+ * the operation {@link BaseThingHandlerFactory#getThingProperties(Thing)} must be overwritten in the concrete thing
+ * handler factory class.
+ * 
+ * @author Thomas Höfer - initial contribution
+ */
+public enum ThingProperty {
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/ThingProperty.java]: I do not like the name ThingProperty - the name suggests much more than it is.
How about something like "DefaultPropertyKeys"?

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/ThingProperty.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
+package org.eclipse.smarthome.core.thing;
+
+import org.eclipse.smarthome.config.core.Configuration;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandlerFactory;
+
+/**
+ * The {@link ThingProperty} enumeration defines all base configuration properties for a {@link Thing}. They
+ * will be injected into the {@link Configuration} of a thing during its instantiation (compare
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/ThingProperty.java]: I would remove lines 2-4 of this JavaDoc. The enum only defines some default keys for properties - nothing more and nothing less.

File: binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/HueThingHandlerFactory.java [hunk 6]
```diff
@@ -116,4 +118,12 @@ protected synchronized void removeHandler(ThingHandler thingHandler) {
             }
         }
     }
+
+    @Override
+    protected Map<ThingProperty, String> getThingProperties(Thing thing) {
```
Comment (kaikreuzer) [inline binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/HueThingHandlerFactory.java]: I would rename the method to getStaticProperties

File: binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueBridgeHandler.java [hunk 7]
```diff
@@ -128,6 +131,14 @@ public void run() {
                                 }
                             }
                         }
+                        
+                        final Config config = fullConfig.getConfig(); 
+                        if (config != null) {
+                            getThing().getConfiguration().put(ThingProperty.SERIAL_NUMBER.keyName,
```
Comment (kaikreuzer) [inline binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueBridgeHandler.java]: You are executing the code on every single polling request. It should instead only be done once on the first successful connection. Or maybe not even then? Maybe only on first connection if the thing does not yet have the properties set?

I've also discussed with @dnobel that we will need a mechanism to tell the system that a thing has been updated and needs to be persisted. This would have to be used here, once it is available.

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java (hunk 1 above)
Comment (tomhoefer) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java]: correct....good catch

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java (hunk 2 above)
Comment (tomhoefer) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java]: ok

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java (hunk 3 above)
Comment (tomhoefer) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java]: sure

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/ThingProperty.java (hunk 5 above)
Comment (tomhoefer) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/ThingProperty.java]: ok

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/ThingProperty.java (hunk 4 above)
Comment (tomhoefer) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/ThingProperty.java]: Below you suggested to rename the operation getThingProperties to getStaticProperties... Then I suggest to rename the enum to StaticPropertyKey...Ok?

File: binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/HueThingHandlerFactory.java (hunk 6 above)
Comment (tomhoefer) [inline binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/HueThingHandlerFactory.java]: ok (cp. comment of renaming enum ThingProperty...)

File: binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueBridgeHandler.java (hunk 7 above)
Comment (tomhoefer) [inline binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueBridgeHandler.java]: But what happens if the firmware version has been changed? My suggestion is that the properties are only filled / updated here if they have not been specified yet or if they have been changed.... ok?

Comment (kaikreuzer) [inline binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueBridgeHandler.java]: ok

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/ThingProperty.java (hunk 4 above)
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/ThingProperty.java]: No, because this method is only in the factory, which can only deal with static values that should always be set. The enum also contains keys for non-static properties, such as the firmware version.

File: docs/sources/architecture/thing-definition.md [hunk 8]
```diff
@@ -154,4 +154,29 @@ The channel group type is defined on the same level as the thing types and chann
...
+
+### Thing Properties
+The configuration descriptions explained in full detail [here](../architecture/configuration.md) are used in order to provide dynamic configuration properties for bindings. In this context *dynamic* means that these configuration properties can be changed at runtime (e.g. because of the IP address of a bridge has been changed). Furthermore these properties are enhanced through configuration descriptions so that the property value can be interpreted in a meaningful manner. However in order to keep the XML based configuration clean there are predefined `ThingProperty`s available that represent properties in a more static manner. In this context *static* means that the value of the property might not be changed (or at least very seldom). In addition to this the type of the property can always be handled as a `java.lang.String` so that no further configuration description is necessary. These *static* properties are defined in the `org.eclipse.smarthome.core.thing.ThingProperty` enumeration which provides the following properties currently:
```
Comment (kaikreuzer) [inline docs/sources/architecture/thing-definition.md]: I would completely remove the "static" stuff here from the doc. What we introduce is simply a list of predefined keys to use for commonly used properties. Some could be set statically by the factory, some dynamically by the handler - but the very same is the case for all other config properties (that do not use the pre-defined keys).

File: docs/sources/architecture/thing-definition.md [hunk 9]
```diff
@@ -154,4 +154,29 @@ The channel group type is defined on the same level as the thing types and chann
...
+
+- Vendor
```
Comment (kaikreuzer) [inline docs/sources/architecture/thing-definition.md]: Didn't we also want to introduce "ID" as a key, which can be used by the inbox?

Comment (tomhoefer) [inline docs/sources/architecture/thing-definition.md]: Yes... But we have removed this from the acceptance criteria of the corresponding US because we have to discuss this in more details.... The required feature could be implemented with US QSDK-2305 (that I will start as soon as this PR is merged)

Comment (dnobel) [inline docs/sources/architecture/thing-definition.md]: Not sure if the ID should be introduced as config property at the thing, too. Imagine your ID is the SERIAL_NUMBER, then you have the information twice at the thing. I think the concept of a identifier should only be introduced at the DiscoveryResult. But we should discuss that again.

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/ThingProperty.java (hunk 4 above)
Comment (tomhoefer) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/ThingProperty.java]: Yes...you are right...I will rename the enum to DefaultPropertyKey

File: binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueBridgeHandler.java [hunk 10]
```diff
@@ -128,6 +132,14 @@ public void run() {
                                 }
                             }
                         }
+                        
+                        final Config config = fullConfig.getConfig(); 
+                        if (config != null) {
+                            storeConfigValue(getThing().getConfiguration(), DefaultPropertyKey.SERIAL_NUMBER,
+                                    config.getMACAddress().replaceAll(":", ""));
```
Comment (kaikreuzer) [inline binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueBridgeHandler.java]: Why do you remove the colons? In the hue app, the bridge identifier is shown including them.

File: binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueLightHandler.java [hunk 11]
```diff
@@ -97,6 +100,13 @@ public void initialize() {
             // note: this call implicitly registers our handler as a listener on the bridge
             if (getHueBridgeHandler() != null) {
                 getThing().setStatus(getBridge().getStatus());
+                getThing().getConfiguration().put(DefaultPropertyKey.SERIAL_NUMBER.name, 
+                        getBridge().getConfiguration().get(SERIAL_NUMBER) + ":" + lightId);
```
Comment (kaikreuzer) [inline binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueLightHandler.java]: Why do you add the light id? This should imho not be part of the serial number?

File: binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/HueThingHandlerFactory.java [hunk 12]
```diff
@@ -116,4 +118,12 @@ protected synchronized void removeHandler(ThingHandler thingHandler) {
             }
         }
     }
+
+    @Override
+    protected Map<DefaultPropertyKey, String> getStaticProperties(Thing thing) {
+        Map<DefaultPropertyKey, String> staticProperties = super.getStaticProperties(thing);
+        staticProperties.put(DefaultPropertyKey.VENDOR, "Philips");
```
Comment (kaikreuzer) [inline binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/HueThingHandlerFactory.java]: This is not correct - the hue binding also supports bulbs from Osram and GE and potentially any other bulb that registers as a "Generic_ZLL" type.

File: binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/internal/YahooWeatherHandlerFactory.java [hunk 13]
```diff
@@ -44,4 +48,11 @@ protected ThingHandler createHandler(Thing thing) {
 
         return null;
     }
+
+    @Override
+    protected Map<DefaultPropertyKey, String> getStaticProperties(Thing thing) {
+        Map<DefaultPropertyKey, String> staticProperties = super.getStaticProperties(thing);
+        staticProperties.put(DefaultPropertyKey.VENDOR, "Yahoo");
```
Comment (kaikreuzer) [inline binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/internal/YahooWeatherHandlerFactory.java]: A vendor property imho does not make sense for a web service. Nobody "bought" it.

File: binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/HueThingHandlerFactory.java [hunk 14]
```diff
@@ -116,4 +118,12 @@ protected synchronized void removeHandler(ThingHandler thingHandler) {
...
+        staticProperties.put(DefaultPropertyKey.MODEL, thing.getThingTypeUID().getId());
```
Comment (kaikreuzer) [inline binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/HueThingHandlerFactory.java]: This is only correct for the ones that are not "ZLL_Light" or "bridge" type.

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java [hunk 15]
```diff
@@ -233,18 +234,32 @@ public Thing createThing(ThingTypeUID thingTypeUID, Configuration configuration,
...
+        return configDescriptionRegistryServiceTracker.getService();
     }
 
+    /**
+     * Retrieves the {@link DefaultPropertyKey}s for the given {@link Thing}. Sub-classes should overwrite this operation
+     * in order to provide these kind of properties for the things to be created.
+     * 
+     * @param thing the {@link Thing} for which the {@link DefaultPropertyKey}s are to be retrieved
+     */
+    protected Map<DefaultPropertyKey, String> getStaticProperties(Thing thing) {
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java]: This should be a Map<String, String> as static properties might not all have a default key.

File: docs/sources/architecture/thing-definition.md [hunk 16]
```diff
@@ -154,4 +154,50 @@ The channel group type is defined on the same level as the thing types and chann
...
+### Default Thing Properties
+The configuration descriptions explained in full detail [here](../architecture/configuration.md) are used in order to provide configuration properties for bindings. These properties are enhanced through configuration descriptions so that the property value can be interpreted in a meaningful manner. However in order to keep the XML based configuration clean there are predefined default thing properties available. The type of these properties can always be handled as a `java.lang.String` so that no further configuration description is necessary. The keys for these properties are defined in the `org.eclipse.smarthome.core.thing.DefaultPropertyKey` enumeration which provides the following property keys:
+
+- Vendor
+- Model
+- Serial Number
+- Hardware Version
+- Firmware Version
+
+To use these properties the operation `getStaticProperties` of the `BaseThingHandlerFactory` class can be overwritten in the concrete thing handler factory class. The following code snippet shows how these thing properties are injected into the thing configuration.
```
Comment (kaikreuzer) [inline docs/sources/architecture/thing-definition.md]: This method is unrelated to the default property keys, so this part should imho be removed from this paragraph.

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/DefaultPropertyKey.java [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+* Copyright (c) 2014-2015 openHAB UG (haftungsbeschraenkt) and others.
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Eclipse Public License v1.0
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v10.html
+*/
+package org.eclipse.smarthome.core.thing;
+
+/**
+ * The {@link DefaultPropertyKey} enumeration defines all keys for default properties of a {@link Thing}.
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/DefaultPropertyKey.java]: Reading this after reviewing your code I think we might better rename this to "CommonPropertyKey" or so. The word "default" somehow implies that every thing has to have these properties, which is not true. We just "suggest" these keys IF a thing has a property of a certain "common" meaning.

File: binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueBridgeHandler.java (hunk 10 above)
Comment (tomhoefer) [inline binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueBridgeHandler.java]: Because this was part of the acceptance criteria defined by the PO. I just recognize one bug: The acceptance criteria defines among others the following:
id=("mac" address of bridge without : and fffe between) from Philips REST API (e.g. "001788fffe169b87"). This is the id also returned by Network UPnP call at https://www.meethue.com/api/nupnp
serialNumber="<serial-number-from-device>" (same is id)

So I have forgotten the fffe in between... And because of the id parameter has not been fully discussed yet I put it as serial number value...as required...

File: binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueLightHandler.java (hunk 11 above)
Comment (tomhoefer) [inline binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueLightHandler.java]: This was also a requirement by the PO

File: binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueBridgeHandler.java (hunk 10 above)
Comment (kaikreuzer) [inline binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueBridgeHandler.java]: Btw, on the sticker on the bridge, the serial is also given with colons. So anywhere where it is presented to the user, it has colons - should hence be the same here.

File: binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueBridgeHandler.java [hunk 18]
```diff
@@ -153,6 +166,13 @@ public void run() {
             }
         }
 
+        private void storeConfigValue(Configuration configuration, CommonPropertyKey key, String newValue) {
+            String existingPropertyValue = (String) configuration.get(key.name);
+            if (existingPropertyValue == null || !existingPropertyValue.equals(newValue)) {
```
Comment (kaikreuzer) [inline binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueBridgeHandler.java]: maybe also add a nullcheck for newValue, just in case?

File: binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueBridgeHandler.java [hunk 19]
```diff
@@ -220,6 +240,9 @@ public void dispose() {
     public void initialize() {
         logger.debug("Initializing hue bridge handler.");
 
+        getConfig().put(CommonPropertyKey.VENDOR.name, HueBindingConstants.VENDOR_PHILIPS);
+        getConfig().put(CommonPropertyKey.MODEL.name, THING_TYPE_BRIDGE.getId());
```
Comment (kaikreuzer) [inline binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueBridgeHandler.java]: Didn't we discuss that "bridge" is not a valid model name and thus this should better not be set?

File: docs/sources/architecture/thing-definition.md [hunk 20]
```diff
@@ -154,4 +154,34 @@ The channel group type is defined on the same level as the thing types and chann
 </channel-group-type> 
 ```
 
-When a thing will be created for a thing type with channel groups, the channel UID will contain the group ID in the last segment divided by a hash (#). If an Item should be linked to channel within a group, the channel UID would be `binding:multiChannelSwitchActor:myDevice:switchActor1#switch` for the XML example before.
\ No newline at end of file
+When a thing will be created for a thing type with channel groups, the channel UID will contain the group ID in the last segment divided by a hash (#). If an Item should be linked to channel within a group, the channel UID would be `binding:multiChannelSwitchActor:myDevice:switchActor1#switch` for the XML example before.
+
+### Common Thing Properties
```
Comment (kaikreuzer) [inline docs/sources/architecture/thing-definition.md]: Throughout this paragraph is should be made clear that this is about predefining KEYs (not properties) that people are supposed to use, if possible.
It is not clear how pre-defining keys leads to a "clean XML".

File: docs/sources/architecture/thing-definition.md [hunk 21]
```diff
@@ -154,4 +154,34 @@ The channel group type is defined on the same level as the thing types and chann
...
+The configuration descriptions explained in full detail [here](../architecture/configuration.md) are used in order to provide configuration properties for bindings. These properties are enhanced through configuration descriptions so that the property value can be interpreted in a meaningful manner. However in order to keep the XML based configuration clean there are predefined common thing properties available. The keys for these properties are defined in the `org.eclipse.smarthome.core.thing.CommonPropertyKey` enumeration which provides the following property keys:
+
+- Vendor
+- Model
+- Serial Number
+- Hardware Version
+- Firmware Version
+
+The common thing property value can either be stored in the `createThing` operation of the concrete thing handler factory class or in the concrete thing handler class during the communication with the thing, e.g. during the execution of the `initialize` operation:
```
Comment (kaikreuzer) [inline docs/sources/architecture/thing-definition.md]: Again: All the following is not specific to properties that use the pre-defined keys, but for any property in general - so it should rather be explained in a separate section. But looking at the code, I have the impression that the concept for those properties is not yet finalized - imho it makes the code pretty ugly. I'd suggest to merge this PR only once this concept also takes the XML into account. This could imho be closely related to https://www.eclipse.org/forums/index.php/t/1015656/.

Comment (dnobel) [inline docs/sources/architecture/thing-definition.md]: How should the concept take XML into account? As we decided to not have separate thing properties, but use configuration values instead, the developer can specify configuration parameter with default value and readonly flag. So we can just document to use the XML for properties like the vendor or model, which should be know at compile-time.

Comment (kaikreuzer) [inline docs/sources/architecture/thing-definition.md]: Right, so if this is our suggestion, we should do it this way - and not introduce a lot of Java code incl. OSRAM-specific constants etc.
The other option is to use the "binding specific XML information" that I referenced and put in some "hints" there for setting the vendor - e.g. only if it is not Philips. In the handlerfactory there could then be a single line that sets the vendor property based on this information.

Comment (dnobel) [inline docs/sources/architecture/thing-definition.md]: Well, as the OSRAM Bulb is specified as a separate thing type, it would be sufficient to specify vendor and model as configuration values just in the existing XML. But "binding specific XML" would definitely help for other bindings.

I forgot that the label property is mandatory in the XML. So if a developer wants to define a "static property" via XML it needs at least the following lines:

> <parameter name="VENDOR" type="text" readyOnly="true">
>    <label>Vendor</label>
>    <default>Philips</default>
> </parameter>

Originally we thought it would be easier to add one line of Java Code like `properties.put(Constant.VENDOR, "Philips")`, as specifying 4 lines of XML. This was the reason for the protected Java method in the BaseThingHandlerFactory.

Comment (kaikreuzer) [inline docs/sources/architecture/thing-definition.md]: > So if a developer wants to define a "static property" via XML it needs at least the following lines

Not really too nice. I start seeing your point about better having read-only properties as something separate. Especially as we are calling them parameters here, not even properties...

> Originally we thought it would be easier to add one line of Java Code like

This makes sense if it is always the same. But as we saw for this binding, there are different possibilities for different things and this makes it ugly when doing it through code.

File: binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueBridgeHandler.java (hunk 18 above)
Comment (tomhoefer) [inline binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueBridgeHandler.java]: ok

File: binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueBridgeHandler.java (hunk 19 above)
Comment (tomhoefer) [inline binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueBridgeHandler.java]: yes...you are right....will remove this

File: docs/sources/architecture/thing-definition.md (hunk 20 above)
Comment (tomhoefer) [inline docs/sources/architecture/thing-definition.md]: ok...will rewrite this once more

File: docs/sources/architecture/thing-definition.md (hunk 21 above)
Comment (tomhoefer) [inline docs/sources/architecture/thing-definition.md]: My suggestion:

-The vendor will become part of the XML configuration.
-Serial numer and firmware version is set in Java Code if possible
-Model is set for lamps in java code, but it is not set for the bridge and zll lights

**Comment by dnobel** [comment]
As agreed, we will close this PR.
